### ISSUE TITLE
Class examples needed

### ISSUE THREAD

**Issue by monk-e-boy** [godotengine/godot-docs#3772]
**Your Godot version:**
3.2.2

**Issue description:**

Class page is a bit rubbish. It's really hard to figure out how to make a class. Classes are pretty fundimental and there should be a super clear example before the complex documentation

This screenshot from a random YouTube video explains classes in 10 seconds. The links below are useless.

![image](https://user-images.githubusercontent.com/17924807/87201838-519a5180-c2f7-11ea-88d5-edc772f61884.png)

None of us are coming to Godot as noobs, we know how to code. We just want to see how to create a class and move along with our lives. Simple examples are such a time saver. Better if we can copy and paste the code :)

**URL to the documentation page (if already existing):**

https://docs.godotengine.org/en/stable/development/cpp/object_class.html
https://docs.godotengine.org/en/stable/getting_started/workflow/best_practices/what_are_godot_classes.html

**Comment by akien-mga** [comment]
Well you're an expert programmer, but you can also start with the basics: https://docs.godotengine.org/en/latest/getting_started/scripting/gdscript/gdscript_basics.html

**Comment by monk-e-boy** [comment]
You talk down to me like I don't know how to code. I actually teach coding at degree level... so I do know how to code. I'm just saying the documentation needs work.

Maybe split that page up into sections? That make sense? Rather than a huge glob of every idea under the sun.

And it's not linked from the class page, so .... ? ... how would I find it?

I think you need to start with inner classes, then talk about how to move them into separate files and load them. 

![image](https://user-images.githubusercontent.com/17924807/87203273-249b6e00-c2fa-11ea-8485-5d145cede47f.png)

The first thing it says is "inside a class file" ... but doesn't explain what that means.

**Comment by Calinou** [comment]
There should definitely be an example of instancing classes from a file in [this page](https://docs.godotengine.org/en/stable/getting_started/workflow/best_practices/what_are_godot_classes.html). The GDScript basics should remain basic (as its name implies), so it's better to just link to that page instead.

**Comment by monk-e-boy** [comment]
![image](https://user-images.githubusercontent.com/17924807/87203815-4fd28d00-c2fb-11ea-98d9-b4a34d631df8.png)

I think this may need a bit of an edit. The text seems to imply that the code is saved in a file called character.gd, then the code seems to include it's self? Or is the capitalization of Character.gd, and character.gd important?

**Comment by monk-e-boy** [comment]
> PS: In the long term, it'd probably be worth adding a GDScript page to [Learn X in Y minutes](https://learnxinyminutes.com/) 🙂

Yes, this would be a good idea. I just wanted to slap a little class in my MeshInstance to help manage some of the 3D objects I was making. That led me down this rabbit hole :) lol. Turns out that the "inner class" was all I needed and it is 90% identical to Python anyway.

You could just link to this guy: https://www.youtube.com/watch?v=m4fz5qA8-Cc     it's a nice quick intro and shows how to organise classes under a Global "namespace" (not sure what the Godot terminology for this is)

**Comment by willnationsdev** [comment]
No it isn't important. The `Character.new()`, in that example, could easily have been replaced with `get_script().new()`. In my head, I think it's more that those segments are supposed to be two different files. One that defines Character and one that loads and instantiates it. So yeah, it could definitely use some editing.

**Comment by monk-e-boy** [comment]
![image](https://user-images.githubusercontent.com/17924807/87205800-2700c680-c300-11ea-8e54-f7b7cddd0fc9.png)

How would I access variable a from inside my class? I can't find an answer to that.

**Comment by Calinou** [comment]
@monk-e-boy I think it's just `.b`, as this is the current syntax to access parent properties/methods.

Godot 4.0 will add a `super` keyword, so it'll be `super.b` there instead.

**Comment by monk-e-boy** [comment]
Did you mean `.a` ? This doesn't work

![image](https://user-images.githubusercontent.com/17924807/87206603-02a5e980-c302-11ea-8dd0-4f32f143140e.png)

**Comment by monk-e-boy** [comment]
> No it isn't important. The `Character.new()`, in that example, could easily have been replaced with `get_script().new()`. In my head, I think it's more that those segments are supposed to be two different files. One that defines Character and one that loads and instantiates it. So yeah, it could definitely use some editing.

This is an interesting point. I think the docs need to talk about what `.new()` is doing vs `._init()` vs `.new(a, b, c)` vs `._init(a, b, c)`

Only because the first thing you want to do with an object is populate it, then start calling methods to manipulate the internal state.

**Comment by willnationsdev** [comment]
@monk-e-boy I am not 100% sure if there is any special allowance for class Bob accessing `a`. As far as I know, GDScript treats inner classes as if they were entirely separate Scripts, wholly distinct and separate from the outer class, but that they are also embedded resources to some extent in that the Script ends up with a constant with the name of the class that then references the internal embedded resource instance. In this way, they can be thought of as embedded resources inside a scene file, ones which have no dedicated file of their own (though not quite the same way since there isn't any sort of strict dependency tracking of Script inner classes like there are for other embedded resources).

As such, I do not believe it is possible for Bob to access any properties associated with the outer class, for it has no reference to such a class that I know of. The `.property` and `.method()` notations that Calinou referenced are for accessing content from a base class that has been inherited from, not from an outer class.

**Comment by Calinou** [comment]
> This is an interesting point. I think the docs need to talk about what .new() is doing vs ._init() vs .new(a, b, c) vs ._init(a, b, c)

`new()` calls the `_init()` method defined in the script. We considered renaming `_init()` to `new()` to make this more explicit, but this was eventually rejected.
